File: src/main/java/com/destino/projeto_destino/services/auth/AuthenticationService.java
package com.destino.projeto_destino.services.auth;


import com.destino.projeto_destino.dto.auth.login.LoginResponseDTO;
import com.destino.projeto_destino.dto.auth.login.LoginUsuarioDto;
import com.destino.projeto_destino.dto.auth.registro.RegistrationResponseDto;
import com.destino.projeto_destino.dto.auth.registro.RegistroDto;
import com.destino.projeto_destino.model.usuario.SessionToken;
import com.destino.projeto_destino.model.usuario.Usuario;
import com.destino.projeto_destino.repository.usuario.SessionRepository;
import com.destino.projeto_destino.repository.usuario.UsuarioRepository;
import com.destino.projeto_destino.util.model.usuario.Cpf.Cpf;
import com.destino.projeto_destino.util.model.usuario.perfil.UserRole;
import com.destino.projeto_destino.util.model.usuario.senha.SenhaValidator;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Base64;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class AuthenticationService {

    private static final String COOKIE_NAME = "session_token";
    private static final long JWT_EXPIRATION_MS = 20 * 60 * 1000; // 20 minutos
    private static final int REFRESH_TOKEN_DAYS = 30; // 30 dias

    private final UsuarioRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtService jwtService;
    private final SessionRepository sessionRepository;

    public RegistrationResponseDto cadastrarUsuario(RegistroDto dados) {
        if (!SenhaValidator.isValid(dados.senha())) {
            return new RegistrationResponseDto(true, "A senha não atende aos requisitos de segurança.");
        }

        if (userRepository.findByCpf(new Cpf(dados.cpf())).isPresent()) {
            return new RegistrationResponseDto(true, "Erro: CPF já cadastrado.");
        }

        if (userRepository.findByEmail(dados.email()).isPresent()) {
            return new RegistrationResponseDto(true, "Erro: Email já está em uso.");
        }

        try {
            Usuario novoUsuario = new Usuario(
                    dados.nome(),
                    dados.sobreNome(),
                    dados.cpf(),
                    dados.email(),
                    dados.telefone(),
                    passwordEncoder.encode(dados.senha()),
                    UserRole.USUARIO,
                    false
            );

            userRepository.save(novoUsuario);
            return new RegistrationResponseDto(false, "Usuário cadastrado com sucesso!");

        } catch (Exception e) {
            return new RegistrationResponseDto(true, "Erro interno ao cadastrar: " + e.getMessage());
        }
    }

    public LoginResponseDTO realizarLogin(LoginUsuarioDto dadosLogin, HttpServletResponse response) {
        try {
            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(dadosLogin.email(), dadosLogin.senha())
            );

            Usuario usuario = (Usuario) authentication.getPrincipal();

            String jwtToken = jwtService.generateToken(usuario, JWT_EXPIRATION_MS);
            String refreshToken = gerarRefreshToken();

            long validadeEmSegundos = Instant.now().plus(REFRESH_TOKEN_DAYS, ChronoUnit.DAYS).getEpochSecond();

            salvarSessaoNoBanco(usuario, refreshToken, validadeEmSegundos);

            criarCookieSessao(response, refreshToken, validadeEmSegundos);

            return new LoginResponseDTO(
                    false,
                    "Login realizado com sucesso",
                    new LoginResponseDTO.UserInfo(
                            usuario.getId(),
                            usuario.getNome() + " " + usuario.getSobreNome(),
                            usuario.getEmail(),
                            usuario.getPerfil().toString(),
                            jwtToken,
                            JWT_EXPIRATION_MS / 1000
                    )
            );

        } catch (BadCredentialsException e) {
            return new LoginResponseDTO(true, "Email ou senha incorretos.", null);
        } catch (Exception e) {
            return new LoginResponseDTO(true, "Erro ao realizar login: " + e.getMessage(), null);
        }
    }

    private String gerarRefreshToken() {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[64];
        random.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }

    private void salvarSessaoNoBanco(Usuario usuario, String token, long validade) {
        sessionRepository.deleteUserSessionToken(usuario.getId());
        SessionToken sessionToken = new SessionToken(token, validade, usuario);
        sessionRepository.save(sessionToken);
    }

    // --- Métodos Auxiliares ---

    private void criarCookieSessao(HttpServletResponse response, String token, long maxAgeSeconds) {
        long duracaoCookie = maxAgeSeconds - Instant.now().getEpochSecond();

        ResponseCookie cookie = ResponseCookie.from(COOKIE_NAME, token)
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(duracaoCookie)
                .sameSite("None")
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    }

    public LoginResponseDTO renovarToken(String tokenSessao, HttpServletResponse response) {
        Optional<SessionToken> sessionTokenOpt = sessionRepository.findByToken(tokenSessao);

        if (sessionTokenOpt.isEmpty()) {
            limparCookie(response);
            return new LoginResponseDTO(true, "Sessão não encontrada.", null);
        }

        SessionToken sessao = sessionTokenOpt.get();
        Usuario usuario = sessao.getUsuario();

        if (Instant.now().getEpochSecond() > sessao.getValidade()) {
            limparCookie(response);
            sessionRepository.delete(sessao);
            return new LoginResponseDTO(true, "Sessão expirada. Faça login novamente.", null);
        }

        // Gera novo JWT de 20 min
        String novoJwt = jwtService.generateToken(usuario, JWT_EXPIRATION_MS);

        // Monta a resposta com a estrutura correta
        return new LoginResponseDTO(
                false,
                "Token renovado com sucesso.",
                new LoginResponseDTO.UserInfo(
                        usuario.getId(),
                        usuario.getNome() + " " + usuario.getSobreNome(),
                        usuario.getEmail(),
                        usuario.getPerfil().toString(),
                        novoJwt,
                        JWT_EXPIRATION_MS / 1000
                )
        );
    }

    private void limparCookie(HttpServletResponse response) {
        ResponseCookie cookie = ResponseCookie.from(COOKIE_NAME, "")
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(0)
                .sameSite("None")
                .build();

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
    }

    @Transactional
    public void realizarLogout(String tokenSessao, HttpServletResponse response) {
        if (tokenSessao != null) {
            Optional<SessionToken> sessao = sessionRepository.findByToken(tokenSessao);
            sessao.ifPresent(sessionToken -> sessionRepository.deleteUserSessionToken(sessionToken.getUsuario().getId()));
        }
        limparCookie(response);
    }
}

File: src/main/kotlin/com/fatec/destino/services/auth/AuthenticationService.kt
package com.fatec.destino.services.auth

import com.fatec.destino.dto.auth.login.LoginResponseDTO
import com.fatec.destino.dto.auth.login.LoginUsuarioDto
import com.fatec.destino.dto.auth.registro.UsuarioRegistrationResponseDto
import com.fatec.destino.dto.auth.registro.UsuarioRegistroDTO
import com.fatec.destino.model.usuario.SessionToken
import com.fatec.destino.model.usuario.Usuario
import com.fatec.destino.repository.auth.SessionRepository
import com.fatec.destino.repository.usuario.UsuarioRepository
import com.fatec.destino.util.model.usuario.senha.SenhaValidator
import jakarta.servlet.http.HttpServletResponse
import jakarta.transaction.Transactional
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseCookie
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.security.SecureRandom
import kotlin.io.encoding.Base64
import kotlin.time.Clock
import kotlin.time.Duration.Companion.days


@Service
class AuthenticationService(
    private val userRepository: UsuarioRepository,
    private val passwordEncoder: PasswordEncoder,
    private val authenticationManager: AuthenticationManager,
    private val jwtService: JwtService,
    private val sessionRepository: SessionRepository
) {
    companion object {
        private const val COOKIE_NAME = "session_token"
        private const val JWT_EXPIRATION_MS = 20L * 60 * 1000 // 20 minutos
        private const val REFRESH_TOKEN_DAYS = 30L
    }

    fun cadastrarUsuario(dados: UsuarioRegistroDTO): UsuarioRegistrationResponseDto {
        if (!SenhaValidator.isValid(dados.senha)) {
            return UsuarioRegistrationResponseDto(true, "A senha não atende aos requisitos de segurança.")
        }

        if (userRepository.findByCpf(dados.cpf) == null) {
            return UsuarioRegistrationResponseDto(true, "Erro: CPF já cadastrado.")
        }

        if (userRepository.findByEmail(dados.email) == null) {
            return UsuarioRegistrationResponseDto(true, "Erro: Email já está em uso.")
        }

        try {
            val novoUsuario = Usuario(
                nome = dados.nome,
                sobreNome = dados.sobreNome,
                cpf = dados.cpf,
                email = dados.email,
                telefone = dados.telefone,
                senha = passwordEncoder.encode(dados.senha)!!
            )

            userRepository.save<Usuario>(novoUsuario)
            return UsuarioRegistrationResponseDto(false, "Usuário cadastrado com sucesso!")
        } catch (e: Exception) {
            return UsuarioRegistrationResponseDto(true, "Erro interno ao cadastrar: " + e.message)
        }
    }

    fun realizarLogin(dadosLogin: LoginUsuarioDto, response: HttpServletResponse): LoginResponseDTO {
        try {
            val authentication: Authentication = authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(dadosLogin.email, dadosLogin.senha)
            )

            val usuario = authentication.principal as Usuario

            val jwtToken = jwtService.generateToken(usuario, JWT_EXPIRATION_MS)
            val refreshToken = gerarRefreshToken()

            val validadeEmSegundos: Long = Clock.System.now().plus(REFRESH_TOKEN_DAYS.days).epochSeconds

            salvarSessaoNoBanco(usuario, refreshToken, validadeEmSegundos)

            criarCookieSessao(response, refreshToken, validadeEmSegundos)

            return LoginResponseDTO(
                false,
                "Login realizado com sucesso",
                LoginResponseDTO.UserInfo(
                    usuario.id,
                    usuario.nome + " " + usuario.sobreNome,
                    usuario.email,
                    usuario.perfil.toString(),
                    jwtToken,
                    JWT_EXPIRATION_MS / 1000
                )
            )
        } catch (_: BadCredentialsException) {
            return LoginResponseDTO(true, "Email ou senha incorretos.", null)
        } catch (e: Exception) {
            return LoginResponseDTO(true, "Erro ao realizar login: " + e.message, null)
        }
    }

    private fun gerarRefreshToken(): String {
        val random: SecureRandom = SecureRandom()
        val bytes = ByteArray(64)
        random.nextBytes(bytes)
        return Base64.UrlSafe.encode(bytes)
    }

    private fun salvarSessaoNoBanco(usuario: Usuario, token: String, validade: Long) {
        sessionRepository.deleteUserSessionToken(usuario.id)
        val sessionToken = SessionToken(token, validade, usuario)
        sessionRepository.save<SessionToken>(sessionToken)
    }

    // --- Métodos Auxiliares ---
    private fun criarCookieSessao(response: HttpServletResponse, token: String?, maxAgeSeconds: Long) {
        val duracaoCookie: Long = maxAgeSeconds - Clock.System.now().epochSeconds

        val cookie = ResponseCookie.from(COOKIE_NAME, token)
            .httpOnly(true)
            .secure(true)
            .path("/")
            .maxAge(duracaoCookie)
            .sameSite("None")
            .build()

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }

    fun renovarToken(tokenSessao: String, response: HttpServletResponse): LoginResponseDTO {
        val session: SessionToken? = sessionRepository.findByToken(tokenSessao)

        if (session == null) {
            limparCookie(response)
            return LoginResponseDTO(true, "Sessão não encontrada.", null)
        }

        val usuario = session.usuario

        if (Clock.System.now().epochSeconds > session.validade) {
            limparCookie(response)
            sessionRepository.delete(session)
            return LoginResponseDTO(true, "Sessão expirada. Faça login novamente.", null)
        }

        // Gera novo JWT de 20 min
        val novoJwt = jwtService.generateToken(usuario, JWT_EXPIRATION_MS)

        // Monta a resposta com a estrutura correta
        return LoginResponseDTO(
            false,
            "Token renovado com sucesso.",
            LoginResponseDTO.UserInfo(
                usuario.id,
                usuario.nome + " " + usuario.sobreNome,
                usuario.email,
                usuario.perfil.toString(),
                novoJwt,
                JWT_EXPIRATION_MS / 1000
            )
        )
    }

    private fun limparCookie(response: HttpServletResponse) {
        val cookie = ResponseCookie.from(COOKIE_NAME, "")
            .httpOnly(true)
            .secure(true)
            .path("/")
            .maxAge(0)
            .sameSite("None")
            .build()

        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }

    @Transactional
    fun realizarLogout(tokenSessao: String, response: HttpServletResponse) {

        val session  = sessionRepository.findByToken(tokenSessao)
        if (session  == null){
            sessionRepository.deleteUserSessionToken(session )
        }

        limparCookie(response)
    }
}
